accounts-db/src/io_uring/file_creator.rs

```diff
@@ -0,0 +1,550 @@
+        &mut self,
+        ring: &mut Completion<FileCreatorState, FileCreatorOp>,
+        res: io::Result<i32>,
+    ) -> io::Result<()>
+    where
+        Self: Sized,
+    {
+        match res {
+            Ok(_) => (),
+            Err(err) if err.kind() == io::ErrorKind::NotFound => {
+                log::warn!(
+                    "retrying file open: {:?}",
+                    CStr::from_bytes_until_nul(&self.path_bytes)
+                );
+                ring.push(FileCreatorOp::Open(Self {
+                    path_bytes: mem::replace(&mut self.path_bytes, Pin::new(vec![])),
+                    dir_handle: self.dir_handle.clone(),
+                    ..*self
+                }));
+                return Ok(());
+            }
+            Err(e) => return Err(e),
+        }
+
+        let backlog = ring.context_mut().mark_file_opened(self.file_key);
+        for (buf, offset, len) in backlog {
+            let op = WriteOp {
+                file_key: self.file_key,
+                offset,
+                buf,
+                write_len: len,
+            };
+            ring.context_mut().submitted_writes_size += len;
+            ring.push(FileCreatorOp::Write(op));
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct CloseOp {
+    file_key: usize,
+}
+
+impl<'a> CloseOp {
+    fn new(file_key: usize) -> Self {
+        Self { file_key }
+    }
+
+    fn entry(&mut self) -> squeue::Entry {
+        opcode::Close::new(types::Fixed(self.file_key as u32)).build()
+    }
+
+    fn complete(
+        &mut self,
+        ring: &mut Completion<FileCreatorState<'a>, FileCreatorOp>,
+        res: io::Result<i32>,
+    ) -> io::Result<()>
+    where
+        Self: Sized,
+    {
+        let _ = res?;
+        ring.context_mut().mark_file_closed(self.file_key);
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct WriteOp {
+    file_key: usize,
+    offset: usize,
+    buf: IoFixedBuffer,
+    write_len: usize,
+}
+
+impl<'a> WriteOp {
+    fn entry(&mut self) -> squeue::Entry {
+        let WriteOp {
+            file_key,
+            offset,
+            buf,
+            write_len,
+        } = self;
+
+        opcode::WriteFixed::new(
+            types::Fixed(*file_key as u32),
+            buf.as_mut_ptr(),
+            *write_len as u32,
+            buf.io_buf_index()
+                .expect("should have a valid fixed buffer"),
+        )
+        .offset(*offset as u64)
+        .ioprio(IO_PRIO_BE_HIGHEST)
+        .build()
+        .flags(squeue::Flags::ASYNC)
+    }
+
+    fn complete(
+        &mut self,
+        ring: &mut Completion<FileCreatorState<'a>, FileCreatorOp>,
+        res: io::Result<i32>,
+    ) -> io::Result<()>
+    where
+        Self: Sized,
+    {
+        let written = res? as usize;
+
+        let WriteOp {
+            file_key,
+            offset: _,
+            ref mut buf,
+            write_len,
+        } = self;
+
+        assert_eq!(written, *write_len, "short write");
```

Comment:
add a comment here that explains that unless specified otherwise, the io uring
worker will retry automatically on EAGAIN

Reply:
done